executor/worker/runtime.go

```diff
@@ -65,82 +67,187 @@ func (s *Scheduler) runImpl() {
 		s.Unlock()
-		if err := worker.Poll(s.ctx); err != nil {
-			s.onWorkerFinished(worker, err)
+		if err := task.Poll(s.ctx); err != nil {
+			s.onTaskFinished(task, err)
 			continue
 		}
-		// TODO: calculate workload
 
 		s.Lock()
-		s.queue.PushBack(worker)
+		s.queue.PushBack(task)
 		s.Unlock()
 	}
 }
 
-func NewRuntime(ctx context.Context) *Runtime {
+func NewRuntime(ctx context.Context, capacity int64) *Runtime {
 	rt := &Runtime{
-		ctx:           ctx,
-		closingWorker: make(chan Runnable, 1024),
-		initingWorker: make(chan Runnable, 1024),
-		scheduler:     Scheduler{ctx: ctx, queue: deque.NewDeque()},
+		closingTasks: make(chan *internal.RunnableContainer, capacity),
+		initingTasks: make(chan *internal.RunnableContainer, capacity),
+		scheduler:    Scheduler{ctx: ctx, queue: deque.NewDeque()},
+		capacity:     capacity,
 	}
-	rt.scheduler.onWorkerFinished = rt.onWorkerFinish
+	rt.scheduler.onTaskFinished = rt.onTaskFinished
 	return rt
 }
 
 type Runtime struct {
-	workerList sync.Map // map[lib.WorkerID]lib.Worker
+	taskList sync.Map // stores *internal.RunnableContainer
 	// We should abstract a schedule interface to implement different
-	// schedule algorithm. For now, we assume every worker consume similar
+	// schedule algorithm. For now, we assume every task consume similar
 	// poll time, so we expect go scheduler can produce a fair result.
-	scheduler     Scheduler
-	closingWorker chan Runnable
-	initingWorker chan Runnable
-	ctx           context.Context
+	scheduler    Scheduler
+	closingTasks chan *internal.RunnableContainer
+	initingTasks chan *internal.RunnableContainer
+
+	capacity int64
+
+	// taskNumMu protects taskNum from concurrent writes
+	// Reading from it does not require locking.
+	taskNumMu sync.Mutex
+	taskNum   atomic.Int64
 }
 
-func (r *Runtime) onWorkerFinish(worker Runnable, err error) {
-	log.L().Warn("Worker has finished",
-		zap.Any("worker-id", worker.ID()),
+func (r *Runtime) onTaskFinished(task *internal.RunnableContainer, err error) {
+	log.L().Warn("Task has finished",
+		zap.Any("worker-id", task.ID()),
 		zap.Error(err))
-	r.closingWorker <- worker
+	task.OnStopped()
+
+	select {
+	case r.closingTasks <- task:
+	default:
+		log.L().Panic("closingTasks is blocked unexpectedly",
+			zap.Int64("capacity", r.capacity),
+			zap.Int64("task-num", r.taskNum.Load()),
+			zap.Int("chan-size", len(r.closingTasks)))
+	}
 }
 
-func (r *Runtime) closeWorker() {
-	for worker := range r.closingWorker {
-		// TODO context and error handling
-		_ = worker.Close(context.Background())
-		r.workerList.Delete(worker.ID())
+func (r *Runtime) closeTask(ctx context.Context) {
+	for {
+		var (
+			task *internal.RunnableContainer
+			ok   bool
+		)
+		select {
+		case <-ctx.Done():
+			return
+		case task, ok = <-r.closingTasks:
+		}
+
+		if !ok {
+			return
+		}
+		// TODO error handling
+		_ = task.Close(ctx)
+		r.taskList.Delete(task.ID())
+
+		r.taskNumMu.Lock()
+		newTaskNum := r.taskNum.Sub(1)
```

Comment:
Isn't atomic sub thread-safe?

Reply:
A lock is used to simplify the implementation of the algorithm below
```
if taskNum < r.capacity {  // L1
    taskNum += 1           // L2
} else {
    // returns error
}
```
Not using a lock will risk other thread's instructions go between L1 and L2.

Reply:
That makes sense.